Return the updated or inserted rate in the response

mikkamp opened the pull request:
### Changes proposed in this Pull Request:

The delete batch endpoint for shipping rates requires a list of ID's. However if we have just inserted new shipping rates (which are auto-saved) we won't know their ID's until we refetch the whole list of shipping rates again.

This PR prevents the refetch from being required by returning the inserted/updated rates within the response.

Note: We do refetch the full shipping rate again from the database, however since this is from the local DB I think this is still acceptable and shouldn't create any additional delays.

### Detailed test instructions:

Send a batch request to update/insert several rates.

`POST https://domain.test/wp-json/wc/gla/mc/shipping/rates/batch`

```json
{
  "rates": [
    {
      "country": "PT",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 80,
      "options": {
        "free_shipping_threshold": "2000"
      }
    },
    {
      "country": "DE",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 50,
      "options": {
        "free_shipping_threshold": "1200"
      }
    },
    {
      "country": "US",
      "method": "flat_rate",
      "currency": "USD",
      "rate": 10
    }
  ]
}
```

Expected response with rates:

```json
{
	"errors": [],
	"success": [
		{
			"status": "success",
			"message": "Successfully added rate for country: \"PT\".",
			"rate": {
				"id": "2",
				"country": "PT",
				"method": "flat_rate",
				"currency": "USD",
				"rate": "80",
				"options": {
					"free_shipping_threshold": 2000
				}
			}
		},
		{
			"status": "success",
			"message": "Successfully added rate for country: \"DE\".",
			"rate": {
				"id": "3",
				"country": "DE",
				"method": "flat_rate",
				"currency": "USD",
				"rate": "50",
				"options": {
					"free_shipping_threshold": 1200
				}
			}
		},
		{
			"status": "success",
			"message": "Successfully added rate for country: \"US\".",
			"rate": {
				"id": "4",
				"country": "US",
				"method": "flat_rate",
				"currency": "USD",
				"rate": "10",
				"options": []
			}
		}
	]
}
```

### Changelog entry

* Tweak - Return the updated or inserted rate in the API response.

mikkamp committed:
Helper function for returning last inserted ID

mikkamp committed:
Return updated or inserted rate in response

mikkamp commented:
Note: There are two failing checks, but these are related to JavaScript code which hasn't been modified in this PR. Since the work for the branch `feature/shipping-rates-api-ui-changes` is still ongoing we can ignore these here.

ecgan commented:
@mikkamp ,

> Note: There are two failing checks, but these are related to JavaScript code which hasn't been modified in this PR. Since the work for the branch `feature/shipping-rates-api-ui-changes` is still ongoing we can ignore these here.

I have changed the PR's base branch to `feature/refactor-shipping-rates-api`, and re-triggered checks. They are now all successful. 🙂 

I will manually merge your branch into the `feature/shipping-rates-api-ui-changes` branch for my testing and consumption on the UI side.

ecgan reviewed (commented):
I tested consuming the API changes here and it is working as expected. Thanks @mikkamp! 🙂  👍  Will leave the PHP code review to others in the team.

jorgemd24 commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -186,14 +186,11 @@ protected function get_create_rate_callback(): callable {
+					$rate_id = $existing_query->get_results()[0]['id'];
+					$update_query->update( $data, [ 'id' => $rate_id ] );
 				} else {
 					$update_query->insert( $data );
```
❓ 

This is not introduced in this PR but the name $update_query is a bit confusing as it is inserting instead of updating. I think it would make more sense something like $shipping_rate_query->update or $shipping_rate_query->insert.

mikkamp replied:
Noticed that as well, I can rename that to make it a bit clearer.

jorgemd24 commented on src/API/Site/Controllers/MerchantCenter/ShippingRateController.php:
```diff
@@ -186,14 +186,11 @@ protected function get_create_rate_callback(): callable {
+					$update_query->update( $data, [ 'id' => $rate_id ] );
 				} else {
 					$update_query->insert( $data );
+					$rate_id = $update_query->last_insert_id();
```
❓ 

I guess that the probability that another plugin or function inserts data at the same time is low and therefore it is safe to use the last_insert_id?

mikkamp replied:
As far as I can see, that wouldn't be possible, because some other code would be running in a separate PHP process which would have a different instance of $wpdb. The $wpdb->insert_id is saved after the insert call so it's not retrieved from DB at a later state.

If we want we can store the last insert ID in the Query class (just in case they aren't called in succession). But I indicated in the function "Must be called after insert". Do you think I should stress that it should be called right after, or do you think it's better to add a safeguard by storing a copy of it in the Query class?

mikkamp replied:
I went with storing a local copy as an extra precaution, see 84b2d24

jorgemd24 reviewed (commented):
I have tested the endpoint `https://domain.test/wp-json/wc/gla/mc/shipping/rates/batch` and it is working as expected: 

✅  I can update & insert data
✅  The response contains the Shipping Rate object. 

I have just left some comments about the naming of one of the variables and about the last_insert_id function.

mikkamp committed:
Rename query variable

The query can be used for update or insert, so calling it
`$update_query` is confusing.

mikkamp committed:
Store a copy of the last inserted ID

This is an extra precaution in case the last inserted ID is not fetched
right away after the insert.

jorgemd24 reviewed (approved):
Thanks, @mikkamp for the changes. LGTM 👍
